### PR TITLE
Begin controller layer TypeScript conversion

### PR DESCRIPTION
**Pull request by bitpshr**
This PR is a proof-of-concept and can serve as a starting point for a conversion of the MetaMask controller layer to TypeScript. A new `BaseController` class was added and the `PreferencesController` and `TokenRatesController` classes were rethought and refactored to extend this new class during the conversion. There's also no longer any reliance on an external library like `obs-store` to manage store subscriptions, as this functionality is provided by the base controller class itself.

**Rationale**

Despite its arguably steeper learning curve compared to JavaScript, TypeScript offers an objectively safer development experience that directly improves the reliability and maintainability of a large codebase:

1. Eliminate unintentionally broken APIs

    TypeScript eliminates the ability to introduce changes to a codebase that break the APIs between or within components. If the signature changes for a specific function in a given class, every use of that function will immediately fail to compile until updated to match the new signature. This same idea applies to every other construct of the language: if the type of a property on an object is changed from `string` to `boolean`, every invalid use of that property will immediately fail to compile, etc. This is one of the most powerful features of a statically typed language for a mission-critical, modular application like MetaMask: it's extremely difficult to unknowingly introduce side effects.

    [Example](https://gist.github.com/bitpshr/5c8f11bdc7467130a5a62e17df350ca8)

1. Share custom types and interfaces

    The ability to describe the shape that code must adhere to allows for the safe consumption and sharing of code without the need for duck typing. An underwhelming but point-driving comparison could be that TypeScript provides an enforced and very advanced form of JSDoc. Common or complex object formats with type-checked members can be shared and guaranteed across a codebase. This eliminates logic paths that only serve to check for the existence or the type of certain properties on an object, which usually account for a large portion of the code paths in a JavaScript application. Codebases become pragmatic and composable.

    [Example](https://gist.github.com/bitpshr/223266bed0eb12b23b777ca5afd36ecf)

1. Eliminate errors caused by misnamed code

    TypeScript exposes a powerful language services API used by most major editors for code introspection features such as member autocompletion. This drastically reduces the chance of misnaming a function or property from an imported module, and in the off chance that a typo or misnaming does occur, the codebase would again immediately fail to compile.

    [Example](https://i.gyazo.com/7995ed6dcc83bc4d01d8283e53fdefd0.gif)

1. Use transpiled language-forward features everywhere

    TypeScript automatically transpiles the latest ECMAScript language features to any desired target gracefully down to ES3. Microsoft directly monitors, supports, and heavily steers the implementation progress of the latest ES2018 specification through its development of TypeScript and adds opt-in support for features as their API churn settles down even before official TC39 finalization. This means that even implementation-heavy features like `async` / `await` are down-emitted without the hassle of manually managing polyfills.

    [Example](https://gist.github.com/bitpshr/cf25c6d2fbee14a79970168473884878)

1. Other general TypeScript features

    * [Generics](https://www.typescriptlang.org/docs/handbook/generics.html)
    * [Member visibility](https://www.typescriptlang.org/docs/handbook/classes.html#public-private-and-protected-modifiers)
    * [Enums](https://www.typescriptlang.org/docs/handbook/enums.html)
    * [Advanced types](https://www.typescriptlang.org/docs/handbook/advanced-types.html) (e.g. union, intersection)
    * Easy [3rd-party typings management](https://github.com/DefinitelyTyped/DefinitelyTyped)
    * [Mixins ](https://www.typescriptlang.org/docs/handbook/mixins.html)
    * Superset of JavaScript, so all JS is valid TS

All of these features of a statically-typed language like TypeScript combine to form codebases consisting of strict contractural relationships between every module in an application. Changes are less intimidating since unnoticed side effects are difficult to introduce, documentation can be generated with less verbose syntax (TSDoc is JSDoc without the types since it can infer them), editors provide autocompletion of both public module members and of local and NPM dependencies, and an upgrade to TypeScript can be incremental as demonstrated in this PR.

A conversion to TypeScript also provides an opportunity to refactor, tighten, and fully test the logic in each controller.

**Implementation notes**

This was work was completed in heavy proof-of-concept fashion; focus was put on rewriting a subset of controllers using better logic and TypeScript and not on the code format or organization itself. If this work is used as a foundation for a true incremental conversion, the following would need to be addressed:

* TypeScript compilation should be handled by gulp (or webpack or ?)
* Proper tasks should be added / updated (watch, dev, dist, etc.)
* Built JS file output would be ignored (kept for now for easy CI integration)
* When a module is converted to TypeScript, so too should it's test
* Potentially rethink our `.editorconfig`, discussion for a different day :)

### CODE REVIEW

File: app/scripts/controllers/ts/BaseController.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
+(function (factory) {
```
Comment (bitpshr) [inline app/scripts/controllers/ts/BaseController.js]: This file built would normally be ignored. It was included in the PR so the gulp build pipeline or CI didn't need to be reconfigured for this POC.

Comment (bitpshr) [inline app/scripts/controllers/ts/BaseController.js]: This file built would normally be ignored. It was included in the PR so the gulp build pipeline or CI didn't need to be reconfigured for this POC.

File: app/scripts/controllers/ts/PreferencesController.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
+(function (factory) {
```
Comment (bitpshr) [inline app/scripts/controllers/ts/PreferencesController.js]: This file built would normally be ignored. It was included in the PR so the gulp build pipeline or CI didn't need to be reconfigured for this POC.

File: app/scripts/controllers/ts/TokenRatesController.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
+var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
```
Comment (bitpshr) [inline app/scripts/controllers/ts/TokenRatesController.js]: This file built would normally be ignored. It was included in the PR so the gulp build pipeline or CI didn't need to be reconfigured for this POC.

File: .eslintignore [hunk 4]
```diff
@@ -3,4 +3,5 @@ test/integration/bundle.js
 test/integration/jquery-3.1.0.min.js
 test/integration/helpers.js
 test/integration/lib/first-time.js
-ui/lib/blockies.js
\ No newline at end of file
+ui/lib/blockies.js
+app/scripts/controllers/ts
```
Comment (bitpshr) [inline .eslintignore]: We'd ideally add TSLint and use that for all TS files. We're ignore compiled JS files here for the sake of this POC.

File: package.json [hunk 5]
```diff
@@ -4,11 +4,11 @@
   "public": false,
   "private": true,
   "scripts": {
-    "start": "gulp dev:extension",
-    "mascara": "gulp dev:mascara & node ./mascara/example/server",
-    "dist": "gulp dist",
+    "start": "npm run build:ts && gulp dev:extension",
```
Comment (bitpshr) [inline package.json]: This was hastily done so TS files are compiled before main NPM tasks, but I was only testing with `start` and `test`. Ideally we'd revisit and greatly simplify our toolchain, or at the very least, compile TypeScript using a gulp task.

**Comment by lazaridiscom** [comment]
[pj.code.refactor] - Refactoring must(!) be done in small incremental steps and in multiple iterations (passes). As an example, one pass can affect only a TypeScript-Converions, the 2nd pass reduces code-duplication. Those passes are clearly separated. Each step is committed, thus other interested developers can follow the refactoring-flow without problems. The high amount of commits can be later reduced (squash before merge).

**Comment by lazaridiscom** [comment]
As for TypeScript:

* Use of TypeScript is a no-brainer: it must be done.
* Personally for me, the rationale is: TypeScript brings JS closer to the King (C++).

As for the Process:
* Introduction should be done in small steps, using  a strict ts-refactoring pass (no code-change, only ts-migration related changes), whilst following the above described basic refactoring procedure.
* identified necessary changes are marked with comments, and applied in a later pass (after ts-refactoring is completed)

And as always: a top-level supervision is needed.

**Review by anthony33lewis — APPROVED** [review]

**Comment by bitpshr** [comment]
Closing in favor of #4491.

**Pull request closed without merging**